scalyr_agent/builtin_monitors/windows_event_log_monitor.py

```diff
@@ -625,18 +636,92 @@ def log_event(self, event):
 
         self._logger.emit_value("EventLog", provider, extra_fields=vals)
 
-        self.__bookmark_lock.acquire()
+        self._bookmark_lock.acquire()
         try:
             if "Channel" in vals:
                 channel = vals["Channel"]
-                bookmark = None
-                if channel not in self.__bookmarks:
-                    self.__bookmarks[channel] = win32evtlog.EvtCreateBookmark(None)
+                if channel not in self._bookmarks:
+                    self._bookmarks[channel] = win32evtlog.EvtCreateBookmark(None)
+                win32evtlog.EvtUpdateBookmark(self._bookmarks[channel], event)
+        finally:
+            self._bookmark_lock.release()
+
+
+class NewJsonApi(NewApi):
+    def __init__(self, config, logger, channels):
+        super(NewJsonApi, self).__init__(config, logger, channels)
 
-                bookmark = self.__bookmarks[channel]
-                win32evtlog.EvtUpdateBookmark(bookmark, event)
+        self._render_context = win32evtlog.EvtCreateRenderContext(
+            win32evtlog.EvtRenderContextSystem
+        )
+
+    def log_event(self, event):
+        values = win32evtlog.EvtRender(
+            event, win32evtlog.EvtRenderEventValues, Context=self._render_context
+        )
+
+        metadata = win32evtlog.EvtOpenPublisherMetadata(
+            values[win32evtlog.EvtSystemProviderName][0]
+        )
+
+        event_json = xmltodict.parse(
+            win32evtlog.EvtFormatMessage(
+                metadata, event, win32evtlog.EvtFormatMessageXml
+            )
+        )
+
+        # Dataset does not currently support arrays/lists, hence must convert to dicts/objects.
+        _convert_json_array_to_object(event_json)
+
+        # If safe, strip off the @ / # prefixes xmltodict uses to prevent conflicts.
+        # Ref: json.dumps(xmltodict.parse('<entry>foo<text>bar</text></entry>'))
+        #        => '{"entry": {"text": "bar", "#text": "foo"}}'
+        _strip_xmltodict_prefixes(event_json)
+
+        # Populate the record here with fields that would normally be added by the log formatter,
+        # this avoids having to unmarshal and remarshal later in the log formatter.
+        # Refer to the use of DummyFormatter in WindowEventLogMonitor.open_metric_log().
+        event_json["timestamp"] = datetime.datetime.utcnow().isoformat(" ")
```

Comment:
Can we actually use timestamp from the WEL event itself (which may be different than the current time, right?

Reply:
Will fix